src/persist/src/indexed/trace.rs

```diff
@@ -26,16 +26,55 @@ use crate::indexed::{BlobTraceBatch, Id, Snapshot};
 use crate::storage::Blob;
 use crate::Data;
 
+/// Metadata describing a [BlobTraceBatch].
+#[derive(Clone, Debug)]
+struct BlobTraceBatchMeta {
+    key: String,
+    desc: Description<u64>,
+    level: u64,
+}
+
 /// A persistent, compacting data structure containing `(Key, Value, Time,
 /// Diff)` entries indexed by `(key, value, time)`.
 ///
+/// We maintain a append-only list of batches that describe updates corresponding
+/// to sorted, contiguous, non-overlapping ranges of times. The `since` frontier
+/// defines time before which we can compact the history of updates (and correspondingly
+/// no longer answer queries about).
+///
+/// We can compact the updates prior to the since frontier physically, by combining
+/// batches representing consecutive intervals into one large batch representing the
+/// union of all intervals, and logically, by forwarding updates at times before
+/// the since frontier to the since frontier.
+///
+/// We also want to achieve a balance between the compactness of the representation
+/// with the computational effort required to maintain the representation. Specifically,
+/// if we have N batches of data already compacted, we don't want every additional
+/// batch to perform O(N) work (i.e. merge with N batches worth of data) in order
+/// to get compacted. Instead, we would like to keep a geometrically decreasing
+/// (when viewed from oldest to most recent) sequence of batches and perform O(N)
+/// work every N calls to append.
+/// Thankfully, we can achieve all of this with a few simple rules:
+///  - Batches are able to be compacted once the since frontier is in advance of all
+///    of the data in the batch.
+///  - All batches are assigned a nonzero compaction level. When a new batch is appended
+///    to the trace it is assigned a compaction level of 1.
+///  - We periodically merge consecutive batches at the same level L representing
+///    time intervals [lo, mid) and [mid, hi) into a single batch representing
+///    all of the updates in [lo, hi) with level L + 1. Once two batches are merged
+///    they are removed from the trace and replaced with the merged batch.
+///  - Perform merges for the oldest batches possible first.
+///
+/// NB: this approach assumes that all batches are roughly uniformly sized when they
+/// are first appended.
+///
 /// Invariants:
 /// - All entries are before some time frontier.
-/// - This acts as an append-only log. Data is added in advancing batches and
-///   logically immutable after that (modulo compactions, which preserve it, but
-///   the ability to read at old times is lost).
-/// - TODO: Explain since and logical compactions.
-/// - TODO: Space usage.
+/// - Batches are sorted by time and represent a sorted, consecutive, non-overlapping
+///   list of time intervals.
+/// - The compaction levels for batches decrease / stay the same when going from oldest
+///   to most recent batch. Specifically, compaction levels never increase.
```

Comment:
nit: "Specifically, compaction levels never increase." is a little confusing because you're talking about iterating from low to high indexes, but an individual batches compaction level does increase. I think we just need a few more words here to clarify